---
title: "Reviewer analyses"
output: html_notebook
---

Reviewer 1 
Does adding in haplogroups change the betas for single variant results?

# Load in ukb data
```{r}
require(dplyr)
require(readr)
options(width=200)
getobj <- function (Rdata){
  require(tools)
  if (tolower(file_ext(Rdata)) == "rds") {
    return(readRDS(Rdata))
  }
  objname <- load(Rdata)
  if (length(objname) > 1) {
    warning(paste("Multiple objects stored in file", Rdata, 
                  "\nReturning only the first object"))
  }
  return(get(objname))
}

# Read in ukb data
# k <- getobj('~/mito_rare-variant/resources/n428k.Subset_removed.nonwhites.nonconsent.exomes_covar_AnnotatedDataFrame.rds')
k <- getobj('~/mtrv/n428k.exomes.nonwhites_nonconsent.removed.AnnotatedDataFrame.rds') # on RAP
k <- as_tibble(k@data) %>% filter(!is.na(arrayCN_PCAodd_m2))

# Read in haplogroups (Code from Dan)
require(data.table)
require(stringr)
haplo<-fread("~/chr26haplogroup1_trim.txt")
tmp<-str_split_fixed(haplo$SampleID,"_",2)
haplo$IID<-as.double(tmp[,1])
haplo<-haplo %>% dplyr::select(IID,Haplogroup)
haplo$Hap2<-substr(haplo$Haplogroup,1,2)
haplo$Hap1<-substr(haplo$Haplogroup,1,1)
haplo<-haplo %>% mutate(Hap=fct_collapse(Hap2,
  L=paste0("L",0:6),
  M=c(paste0("M",1:9),paste0("Q",1:3),paste0("C",1:7),paste0("Z",1:3),"C","Z","Z1","Z3","E1","D","G",paste0("G",1:4),paste0("D",1:6)),
  N=c("N","N5","N9","Y1","Y2","A","A+",paste0("A",1:5),"S3"),
  N1=c("N1","N7",paste0("I",1:6)),
  N2=c("N2",paste0("W",1:8),"W"),
  X=c("X","X2","X3","X4"),
  R=c("P1","P3","P4","P5",paste0("B",1:6),paste0("R",1:8),"R9",paste0("F",1:4)),
  R0=c("R0",paste0("H",1:9),"HV","V+",paste0("V",1:9)),
  U=c("U",paste0("U",1:9),"K",paste0("K",1:3)),
  JT=c("T","JT","T1","T2","T3","J","J1","J2")
  ))
table(haplo$Hap)

k <- k %>% inner_join(haplo, by=c('sample.id'='IID'))

# Read in qc genoptying Batch variable (UKB data field 22000)
qc <- read_table('~/ukb_qc_IDs.txt') %>% select(IID, genotyping.array, Batch)
k <- k %>% inner_join(qc, by=c('sample.id'='IID'))
```

Get variant vectors for genome-wide significant single-variant associations
```{r}
require(SeqArray)
require(SeqVarTools)

genes = c('SAMHD1','SAMHD1','SIRPA','JAK2', 'TWNK', 'TOP3A', 'ACSL1','AHCY', 'PM20D1', 'AFAP1L1')
chrs = c(20, 20, 20, 9, 10, 17, 4, 20, 1, 5)
positions = c(36893060, 36898455, 1895355, 5073770, 100993257, 
              18291048, 184766776, 34280961, 205848004, 149332902)
variants <- tibble(genes, chrs, positions)

# Get variant vectors
geno_variants <- list()
for(i in 1:nrow(variants)){
  message(paste0(variants[i,], sep='\t'))
  chr=variants[i, 'chrs']
  gds <- seqOpen(paste0('~/ukb/UKB\ -mtDNA_2022/rv/resources/gds/UKBexomeQQFE_chr', variants$chrs[i], '.gds'))
  samp.id <- seqGetData(gds, 'sample.id')
  var.id <- seqGetData(gds, 'variant.id')
  pos <- seqGetData(gds, 'position')
  
  seqSetFilter(gds, sample.sel=samp.id %in% k$sample.id)
  seqSetFilter(gds, variant.sel=var.id[which(pos %in% variants$positions[i])])
  
  geno_variant <- altDosage(gds)
  geno_variant <- tibble::rownames_to_column(as.data.frame(geno_variant)) %>% 
    as_tibble() %>% 
    rename(sample.id = rowname)
  
  if(i==2){
    # SAMHD1 at 36898455 has two ALTs - 2nd one is the effect allele, take that:
    geno_variant <- geno_variant %>% select(sample.id, `330236`)
  } else if(i==3){
    # SIRPA at 1895355 has three ALTs - 2nd one is the effect allele, take that:
    geno_variant <- geno_variant %>% select(sample.id, `25483`)
  } else if(i==10){
    # AFAP1L1 at 149332902 has two ALTs - 1st one is the effect allele, take that:
    geno_variant <- geno_variant %>% select(sample.id, `892091`)
  } else if(i==6){
    # TOP3A's alleles are coded differently, change to correct coding
    geno_variant$var1 <- rep(0, nrow(geno_variant))
    geno_variant$var1[which(geno_variant$`429419`==1)] <- 1
    geno_variant$var1[which(geno_variant$`429419`==0)] <- 2
    geno_variant <- geno_variant %>% select(sample.id, var1)
  }

  geno_variant <-
    geno_variant %>% rename(var = colnames(geno_variant)[2])
  geno_variant$sample.id <- as.numeric(geno_variant$sample.id)
  geno_variants[[i]] <- geno_variant
  seqClose(gds)
}
```

Single variants Haplogroups analysis
# run LM adjusting for haplotype
```{r}
lm.vals <- list()
anovas <- list()
for(i in 1:nrow(variants)){
  message(paste0(variants[i,], sep='\t'))
  vals.0 <- rep(NA, 4) # beta, se, p-val, r2
  vals.1 <- rep(NA, 4)
  # names(vals.0) <- c('beta1','se1','pval1','r2-1')
  # names(vals.1) <- c('beta2','se2','pval2','r2-2')
  geno_variant <- geno_variants[[i]]
  
  j <- k %>% inner_join(geno_variant, by='sample.id')
  
  lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j)
  vals.0 <- c(summary(lm0)$coef[2,1], summary(lm0)$coef[2,2], summary(lm0)$coef[2,4], summary(lm0)$r.squared)
  
  lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch + Hap'), data = j)
  vals.1 <- c(summary(lm1)$coef[2,1], summary(lm1)$coef[2,2], summary(lm1)$coef[2,4], summary(lm1)$r.squared)
  anovas[[i]] <- anova(lm0, lm1)
  
  lm.vals[[i]] <- c(vals.0, vals.1)
  names(lm.vals[[i]]) <- c('beta1','se1','pval1','r2-1', 'beta2','se2','pval2','r2-2')
}

lm.vals <- bind_rows(lm.vals)
lm.vals <- lm.vals %>% mutate(r2.diff=`r2-2`-`r2-1`) %>% mutate(beta.diff=beta2 - beta1)

bind_cols(variants, lm.vals) %>% write_tsv('single_variant.haplogroup_analyses.txt')
```

Single variants genotyping batch analysis
# run LM adjusting for chip batch variable
```{r}
lm.vals <- list()
anovas <- list()
for(i in 1:nrow(variants)){
  message(paste0(variants[i,], sep='\t'))
  vals.0 <- rep(NA, 4) # beta, se, p-val, r2
  vals.1 <- rep(NA, 4)
  # names(vals.0) <- c('beta1','se1','pval1','r2-1')
  # names(vals.1) <- c('beta2','se2','pval2','r2-2')
  geno_variant <- geno_variants[[i]]
  
  j <- k %>% inner_join(geno_variant, by='sample.id')
  
  lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j)
  vals.0 <- c(summary(lm0)$coef[2,1], summary(lm0)$coef[2,2], summary(lm0)$coef[2,4], summary(lm0)$r.squared)
  
  lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch + Batch'), data = j)
  vals.1 <- c(summary(lm1)$coef[2,1], summary(lm1)$coef[2,2], summary(lm1)$coef[2,4], summary(lm1)$r.squared)
  anovas[[i]] <- anova(lm0, lm1)
  
  lm.vals[[i]] <- c(vals.0, vals.1)
  names(lm.vals[[i]]) <- c('beta1','se1','pval1','r2-1', 'beta2','se2','pval2','r2-2')
}

lm.vals <- bind_rows(lm.vals)
lm.vals <- lm.vals %>% mutate(r2.diff=`r2-2`-`r2-1`) %>% mutate(beta.diff=beta2 - beta1)

bind_cols(variants, lm.vals) %>% write_tsv('single_variant.genotyping_batch_analyses.txt')
```

Single variants genotyping array analysis
# run LM adjusting for genotpying array variable (UKBB vs UKBL)
```{r}
lm.vals <- list()
anovas <- list()
for(i in 1:nrow(variants)){
  message(paste0(variants[i,], sep='\t'))
  vals.0 <- rep(NA, 4) # beta, se, p-val, r2
  vals.1 <- rep(NA, 4)
  # names(vals.0) <- c('beta1','se1','pval1','r2-1')
  # names(vals.1) <- c('beta2','se2','pval2','r2-2')
  geno_variant <- geno_variants[[i]]
  
  j <- k %>% inner_join(geno_variant, by='sample.id')
  
  lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j)
  vals.0 <- c(summary(lm0)$coef[2,1], summary(lm0)$coef[2,2], summary(lm0)$coef[2,4], summary(lm0)$r.squared)
  
  lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch + genotyping.array'), data = j)
  vals.1 <- c(summary(lm1)$coef[2,1], summary(lm1)$coef[2,2], summary(lm1)$coef[2,4], summary(lm1)$r.squared)
  anovas[[i]] <- anova(lm0, lm1)
  
  lm.vals[[i]] <- c(vals.0, vals.1)
  names(lm.vals[[i]]) <- c('beta1','se1','pval1','r2-1', 'beta2','se2','pval2','r2-2')
}

lm.vals <- bind_rows(lm.vals)
lm.vals <- lm.vals %>% mutate(r2.diff=`r2-2`-`r2-1`) %>% mutate(beta.diff=beta2 - beta1)

bind_cols(variants, lm.vals) %>% write_tsv('single_variant.genotyping_array_analyses.txt')
```

Plot betas of haplogroup analyses
```{r}
require(ggplot2)
require(readr)
require(hrbrthemes)
require(ggrepel)
t <- readr::read_table('~/mito_rare-variant/analyses/single_variant.haplogroup_analyses.txt')
t <- t %>% mutate(abs_beta.diff = abs(beta2 - beta1))

t %>% ggplot(aes(x=beta1, y=beta2, label=genes)) + 
  geom_point() +
  theme_ipsum_rc(axis_text_size = 15) +
  xlab('Beta without Haplogroup Covariate') +
  ylab('Beta with Haplogroup Covariate') +
  labs(title='Betas of single variants with and without Haplogroup') +
  geom_abline(intercept = 0, slope=1, color='red', linetype='dashed') #+
  # ggrepel::geom_text_repel()

cor(t$beta1, t$beta2)
t.test(t$beta1, t$beta2, paired = T)
mean(t$abs_beta.diff)

```

Plot betas of genotyping batch analyses
```{r}
require(ggplot2)
require(readr)
require(hrbrthemes)
require(ggrepel)
t <- readr::read_table('~/mito_rare-variant/analyses/single_variant.genotyping_batch_analyses.txt')
t <- t %>% mutate(abs_beta.diff = abs(beta2 - beta1))

t %>% ggplot(aes(x=beta1, y=beta2, label=genes)) + 
  geom_point() +
  theme_ipsum_rc(axis_title_size = 15) +
  xlab('Beta without Genotyping Chip Batch Covariate') +
  ylab('Beta with Genotyping Chip Batch Covariate') +
  labs(subtitle='Genotyping Chip Batch = UKB Data-Field 22000') +
  labs(title='Betas of single variants\nwith and without Genotyping Chip Batch') +
  geom_abline(intercept = 0, slope=1, color='red', linetype='dashed') #+
  # ggrepel::geom_text_repel()

cor(t$beta1, t$beta2)
t.test(t$beta1, t$beta2, paired = T)
mean(t$abs_beta.diff)


```

Plot betas of genotyping array (UKBB vs. UKBL) analyses
# NOTE: gneotyping array is encapsulated / explained by the chip batch variable, 
# so not necessary to include but for posterity:
```{r}
# > k %>% count(genotyping.array, Batch)
# # A tibble: 106 Ã— 3
#    genotyping.array Batch          n
#    <chr>            <chr>      <int>
#  1 UKBB             Batch_b001  3782
#  2 UKBB             Batch_b002  3905
#  3 UKBB             Batch_b003  3774
#  4 UKBB             Batch_b004  3976
#  5 UKBB             Batch_b005  3794
#  6 UKBB             Batch_b006  3935
#  7 UKBB             Batch_b007  4044
#  8 UKBB             Batch_b008  3968
#  9 UKBB             Batch_b009  3838
# 10 UKBB             Batch_b010  3946
# # â€¦ with 96 more rows

# > k %>% count(Batch)
# # A tibble: 106 Ã— 2
#    Batch          n
#    <chr>      <int>
#  1 Batch_b001  3782
#  2 Batch_b002  3905
#  3 Batch_b003  3774
#  4 Batch_b004  3976
#  5 Batch_b005  3794
#  6 Batch_b006  3935
#  7 Batch_b007  4044
#  8 Batch_b008  3968
#  9 Batch_b009  3838
# 10 Batch_b010  3946
# # â€¦ with 96 more rows

require(ggplot2)
require(readr)
require(hrbrthemes)
require(ggrepel)
t <- readr::read_table('~/mito_rare-variant/analyses/single_variant.genotyping_array_analyses.txt')

t %>% ggplot(aes(x=beta1, y=beta2, label=genes)) + 
  geom_point() +
  theme_ipsum_rc(axis_title_size = 15) +
  xlab('Beta without Genotyping Array Covariate') +
  ylab('Beta with Genotyping Array Covariate') +
  labs(subtitle='Genotyping Array (UKBB vs UKBL) = UKB Data-Field 22000') +
  labs(title='Betas of single variants\nwith and without Genotyping Array') +
  geom_abline(intercept = 0, slope=1, color='red', linetype='dashed') #+
  # ggrepel::geom_text_repel()
```

Single variants blood count outlier analyses
```{r}
blood.counts <- read_table('~/blood_cell_counts_mtDNA-CN_m2_participant.tsv') # on UKB under /rv/resources/
blood.counts <- blood.counts[complete.cases(blood.counts),]
blood.counts <- blood.counts %>% mutate(lWBC=log(leukocyte+1),lRBC=log(rbc+1),lLymph=log(lymphocyte+1),lMono=log(monocyte+1),lNeutrophil=log(neutrophill+1),lEos=log(eosinophill+1),lBaso=log(basophill+1),lNucRBC=log(nuc_rbc+1))

blood.counts <- blood.counts %>%
  mutate(Exclude2=ifelse((lWBC>1.25 & lWBC<3 & lRBC<2 & lRBC>1.4 & platelet<500 & platelet>10 & lLymph>0.10 & lLymph<2 & lMono<0.9 & lNeutrophil>0.75 & lNeutrophil<2.75 & lEos<0.75 & lBaso<0.45),0,1))

k <- k %>% left_join(blood.counts, by=c('sample.id'='IID'))

lm.vals <- list()
# anovas <- list()
for(i in 1:nrow(variants)){
  message(paste0(variants[i,], sep='\t'))
  vals.0 <- rep(NA, 5) # beta, se, p-val, r2, n
  vals.1 <- rep(NA, 5)
  # names(vals.0) <- c('beta1','se1','pval1','r2-1')
  # names(vals.1) <- c('beta2','se2','pval2','r2-2')
  geno_variant <- geno_variants[[i]]
  
  j0 <- k %>% inner_join(geno_variant, by='sample.id')
  j1 <- k %>% inner_join(geno_variant, by='sample.id') %>% filter(Exclude2!=1)
  
  lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0)
  vals.0 <- c(summary(lm0)$coef[2,1], summary(lm0)$coef[2,2], summary(lm0)$coef[2,4], summary(lm0)$r.squared, nrow(model.frame(lm0)))
  
  lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j1)
  vals.1 <- c(summary(lm1)$coef[2,1], summary(lm1)$coef[2,2], summary(lm1)$coef[2,4], summary(lm1)$r.squared, nrow(model.frame(lm1)))
  # anovas[[i]] <- anova(lm0, lm1)
  
  lm.vals[[i]] <- c(vals.0, vals.1)
  names(lm.vals[[i]]) <- c('beta1','se1','pval1','r2-1','sample.size-1', 'beta2','se2','pval2','r2-2','sample.size-2')
}

lm.vals <- bind_rows(lm.vals)
lm.vals <- lm.vals %>% mutate(r2.diff=`r2-2`-`r2-1`) %>% mutate(beta.diff=beta2 - beta1)

bind_cols(variants, lm.vals) %>% write_tsv('single_variant.cell_count_outlier_analyses.txt')
```

```{r}
require(ggplot2)
require(readr)
require(hrbrthemes)
require(ggrepel)
t <- readr::read_table('~/mito_rare-variant/analyses/single_variant.cell_count_outlier_analyses.txt')
t <- t %>% mutate(abs_beta.diff = abs(beta2 - beta1))
t$label <- rep(NA, nrow(t))
t$label[which(t$genes=='JAK2')] <- 'JAK2'
# t$label[which(t$genes=='TWNK')] <- 'TWNK' # not significantly diff (interaction term of TWNK:Exclude2 is > (0.05/11))

t %>% ggplot(aes(x=beta1, y=beta2, label=label)) + 
  geom_point() +
  theme_ipsum_rc(axis_title_size = 15) +
  xlab('Beta before removing Cell Count Outlier Individuals') +
  ylab('Beta after removing Cell Count Outlier Individuals') +
  # labs(subtitle='Genotyping Chip Batch = UKB Data-Field 22000') +
  labs(title='Betas of single variants\nbefore and after removing cell count outlier individuals') +
  geom_abline(intercept = 0, slope=1, color='red', linetype='dashed') +
  ggrepel::geom_label_repel(nudge_x = 0.5)

t.test(t$beta1, t$beta2, paired = T)
# 	Paired t-test
# 
# data:  t$beta1 and t$beta2
# t = 0.432, df = 9, p-value = 0.6759
# alternative hypothesis: true mean difference is not equal to 0
# 95 percent confidence interval:
#  -0.02370507  0.03489598
# sample estimates:
# mean difference 
#     0.005595456 
mean(t$abs_beta.diff)
```

Blood cell counts in regression of JAK2 and TWNK variants
```{r}
blood.counts <- read_table('~/blood_cell_counts_mtDNA-CN_m2_participant.tsv') # on UKB under /rv/resources/
blood.counts <- blood.counts[complete.cases(blood.counts),]
blood.counts <- blood.counts %>% mutate(lWBC=log(leukocyte+1),lRBC=log(rbc+1),lLymph=log(lymphocyte+1),lMono=log(monocyte+1),lNeutrophil=log(neutrophill+1),lEos=log(eosinophill+1),lBaso=log(basophill+1),lNucRBC=log(nuc_rbc+1))

blood.counts <- blood.counts %>%
  mutate(Exclude2=ifelse((lWBC>1.25 & lWBC<3 & lRBC<2 & lRBC>1.4 & platelet<500 & platelet>10 & lLymph>0.10 & lLymph<2 & lMono<0.9 & lNeutrophil>0.75 & lNeutrophil<2.75 & lEos<0.75 & lBaso<0.45),0,1))

k <- k %>% left_join(blood.counts, by=c('sample.id'='IID'))
k$arrayCN_PCAodd_m1[which(is.na(k$arrayCN_PCAodd_m2))] <- NA
```

JAK2 removing blood cell count outliers (testing m2 which already adjusts for cell counts)
```{r}
geno_variant <- geno_variants[[4]] # _JAK2_
j0 <- k %>% inner_join(geno_variant, by='sample.id') # 284 _JAK2_ carriers
j1 <- k %>% inner_join(geno_variant, by='sample.id') %>% filter(Exclude2!=1) # 160 _JAK2_ carriers (meaning 124 are cell count outliers)

lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0) # cell count outliers in
lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j1) # cell count outliers out
# betas are 0.5 (lm0) and 0.4 (lm1) (i.e. before and after removing) pvals drop too

# is the beta difference significant?
lm0.x <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var*Exclude2 + Exclude2 + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0) # cell count outliers interaction term
# interaction term 
# var:Exclude2 -1.466e+00  7.087e-01  -2.068 0.038658 *
# is not signficant when correcting for 11 variants (0.5 / 11)

```

TWNK removing blood cell count outliers (testing m2 which already adjusts for cell counts)
```{r}
geno_variant <- geno_variants[[5]] # _TWNK_
geno_variant %>% count(var)
j0 <- k %>% inner_join(geno_variant, by='sample.id') # 48 _TWNK_ carriers
j1 <- k %>% inner_join(geno_variant, by='sample.id') %>% filter(Exclude2!=1) # 46 _TWNK_ carriers (meaning 2 are cell count outliers)

lm0 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0) # cell count outliers in
lm1 <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j1) # cell count outliers out
# betas are 0.5 (lm0) and 0.4 (lm1) (i.e. before and after removing) pvals drop too

# is the beta difference significant?
lm0.x <- lm(formula=paste0('arrayCN_PCAodd_m2 ~ var*Exclude2 + Exclude2 + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0) # cell count outliers interaction term
# interaction term 
# var:Exclude2 -1.466e+00  7.087e-01  -2.068 0.038658 *
# is not signficant when correcting for 11 variants (0.5 / 11)

```

JAK2 with blood cell counts (testing m1 which doesn't have cell counts adjusted for)
```{r}
# j0 = without cell count outlier removals
# j1 = with cell count outlier removals
j0 <- k %>% inner_join(geno_variant, by='sample.id')
j1 <- k %>% inner_join(geno_variant, by='sample.id') %>% filter(Exclude2!=1)

# Not removing cell count outlier individuals
lm0 <- lm(formula=paste0('arrayCN_PCAodd_m1 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0)
lm0.w_cells <- lm(formula=paste0('arrayCN_PCAodd_m1 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch +
                         platelet + lWBC + lRBC + lLymph + lMono + lNeutrophil + lEos + lBaso + lNucRBC'), data = j0)
# removing cell count outlier individuals
lm1 <- lm(formula=paste0('arrayCN_PCAodd_m1 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j1)
lm1.w_cells <- lm(formula=paste0('arrayCN_PCAodd_m1 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch +
                         platelet + lWBC + lRBC + lLymph + lMono + lNeutrophil + lEos + lBaso + lNucRBC'), data = j1)

# results
lm.vals <- list()
lm.vals[[1]] <- summary(lm0)$coef[2,]
lm.vals[[2]] <- summary(lm0.w_cells)$coef[2,]
lm.vals[[3]] <- summary(lm1)$coef[2,]
lm.vals[[4]] <- summary(lm1.w_cells)$coef[2,]
lm.vals <- bind_rows(lm.vals)

lm.vals <- bind_cols(tibble(Regression=c('All individuals: Null', 'All individuals: Cell Counts Adjusted',
                              'Cell count outlier indivdiuals removed: Null', 'Cell count outlier indivdiuals removed: Cell Counts Adjusted')),
          lm.vals, `Sample Size`=c(nrow(model.frame(lm0)), nrow(model.frame(lm0.w_cells)),
                     nrow(model.frame(lm1)), nrow(model.frame(lm1.w_cells))))

# display table
lm.vals <- read_tsv('~/mito_rare-variant/analyses/jak2.cell_counts_analysis.txt')
lm.vals$Regression <- c('Null','Cell Counts Adjusted','Null','Cell Counts Adjusted')
require(gt)
lm.vals %>% gt() %>% 
  cols_label(Estimate = 'Beta') %>% 
  fmt_number(decimals=2, 
             columns=c('Estimate',`Std. Error`,`t value`)) %>% 
  tab_row_group(label='Cell count outlier indivdiuals removed:', rows=c(3,4)) %>% 
  tab_row_group(label='All Indivdiuals:', rows=c(1,2)) %>% 
  fmt_number(columns = `Sample Size`, sep_mark = ',', decimals = 0) %>% 
  fmt_scientific(columns=c(`Pr(>|t|)`)) %>% 
  tab_header(title = 'mtDNA-CN ~ JAK2.V617F') %>% 
  tab_footnote(footnote = '- Null model: mtDNA-CN ~ JAK2.V617F + age + sex + PCs 1-40 + WES Batch') %>% 
  tab_footnote(footnote = '- Cell Counts Adjusted model: mtDNA-CN ~ JAK2.V617F + age + sex + PCs 1-40 + WES Batch + platelet + lWBC + lRBC + lLymph + lMono + lNeutrophil + lEos + lBaso + lNucRBC')
```

JAK2 and cell counts
```{r}
geno_variant <- geno_variants[[4]] # _JAK2_

j0 <- k %>% inner_join(geno_variant, by='sample.id')

lm0.w_cells <- lm(formula=paste0('arrayCN_PCAodd_m1 ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch +
                         platelet + lWBC + lRBC + lLymph + lMono + lNeutrophil + lEos + lBaso + lNucRBC'), data = j0)
lm0.w_cells <- getobj('~/mito_rare-variant/analyses/JAK2.mtDNA_CN_m1.lm0.w_cells.rds')
lm0.w_cells %>% 
  tbl_regression(include = c("var","age","sex","wes.batch","platelet","lWBC","lRBC",
                             "lLymph","lMono","lNeutrophil","lEos","lBaso","lNucRBC"))

jak2.platelets <- lm(formula=paste0('platelet ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0)
jak2.neutrophills <- lm(formula=paste0('lNeutrophil ~ var + age + sex+',paste('PC',1:40,sep='',collapse='+'),' + wes.batch'), data = j0)


```













